List Panic

### Describe the bug:
When a list displays fewer items than would fill the viewport, it releases unused items without a bounds check

```
panic: runtime error: index out of range [-1]

goroutine 15 [running]:
fyne.io/fyne/widget.(*listRenderer).Layout(0xc000170680, 0x318, 0x22f)
	/Users/stuartscott/Documents/Projects/Go/src/fyne.io/fyne/widget/list.go:127 +0x63b
fyne.io/fyne/widget.(*listRenderer).Refresh(0xc000170680)
	/Users/stuartscott/Documents/Projects/Go/src/fyne.io/fyne/widget/list.go:155 +0x54
fyne.io/fyne/widget.(*BaseWidget).Refresh(0xc00022c480)
	/Users/stuartscott/Documents/Projects/Go/src/fyne.io/fyne/widget/widget.go:137 +0x5c
```

```go
if l.visibleItemCount > 0 && l.list.Length() < l.visibleItemCount && len(l.children) > 0 { //<-- actual bounds check for children
	for i := l.visibleItemCount; i >= l.list.Length(); i-- {                           //<-- expected bounds check for children
		l.itemPool.Release(l.children[len(l.children)-1])
		l.children = l.children[:len(l.children)-1]                                //<-- children modification
	}
}
```

### To Reproduce:
Steps to reproduce the behaviour:
1. Create a dataset with one element
2. Create and display a list using the dataset
3. Add an element to the dataset
4. Call list.Refresh

### Device (please complete the following information):
 - OS: MacOS
 - Version: Catalina
 - Go version: go1.15 darwin/amd64
 - Fyne version: d12ce856aa1e477c8492dd503fc30856caf1a8fa

I think that I am running into the same issue on my end. I am using an empty list where items are added when they are created by the user:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x6a1da9]

goroutine 1 [running, locked to thread]:
fyne.io/fyne/widget.(*listLayout).MinSize(0xc0001c8020, 0x0, 0x0, 0x0, 0x8ad940, 0x2)
        /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/widget/list.go:411 +0x29
fyne.io/fyne.NewContainerWithLayout(0x999d40, 0xc0001c8020, 0x0, 0x0, 0x0, 0x0)
        /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/container.go:138 +0x9b
fyne.io/fyne/widget.(*List).CreateRenderer(0xc0000d2580, 0x0, 0x0)
        /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/widget/list.go:49 +0xbd
fyne.io/fyne/internal/cache.Renderer(0x9a3760, 0xc0000d2580, 0xc0000d2580, 0x0)
        /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/internal/cache/widget.go:29 +0x119
fyne.io/fyne/widget.(*BaseWidget).MinSize(0xc0000d2580, 0x9a2d60, 0xc0000d2580)
        /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/widget/widget.go:84 +0x45
fyne.io/fyne/widget.(*List).MinSize(0xc0000d2580, 0x0, 0x0)
        /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/widget/list.go:60 +0x4c
fyne.io/fyne/layout.(*boxLayout).MinSize(0xc0000b1d7c, 0xc00009bcb0, 0x3, 0x3, 0x87d500, 0xc000099901)
        /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/layout/boxlayout.go:152 +0x1df
fyne.io/fyne.NewContainerWithLayout(0x999b00, 0xc0000b1d7c, 0xc00009bcb0, 0x3, 0x3, 0xc0000d2630)
        /home/jacob/go/pkg/mod/fyne.io/fyne@v1.3.4-0.20200916214832-d12ce856aa1e/container.go:138 +0x9b
main.(*appData).sendTab(0xc00009b680, 0x9a4a60)
        /home/jacob/go/src/github.com/Jacalz/wormhole-gui/send.go:88 +0x45a
main.main()
        /home/jacob/go/src/github.com/Jacalz/wormhole-gui/main.go:39 +0x1e5
```

--- a/widget/list.go
+++ b/widget/list.go
@@ -122,7 +122,7 @@ func (l *listRenderer) Layout(size fyne.Size) {
 			l.children = l.children[:len(l.children)-1]
 		}
 	}
-	if l.visibleItemCount > 0 && l.list.Length() < l.visibleItemCount && len(l.children) > 0 {
+	if l.visibleItemCount > 0 && l.list.Length() < l.visibleItemCount && len(l.children) > l.visibleItemCount-1 {
 		for i := l.visibleItemCount; i >= l.list.Length(); i-- {
 			l.itemPool.Release(l.children[len(l.children)-1])
 			l.children = l.children[:len(l.children)-1]

--- a/widget/list_test.go
+++ b/widget/list_test.go
@@ -15,7 +15,7 @@ import (
 )
 
 func TestNewList(t *testing.T) {
-	list := createList()
+	list := createList(1000)
 
 	template := newListItem(fyne.NewContainerWithLayout(layout.NewHBoxLayout(), NewIcon(theme.DocumentIcon()), NewLabel("Template Object")), nil)
 	firstItemIndex := test.WidgetRenderer(list).(*listRenderer).firstItemIndex
@@ -30,7 +30,7 @@ func TestNewList(t *testing.T) {
 }
 
 func TestList_Resize(t *testing.T) {
-	list := createList()
+	list := createList(1000)
 	w := test.NewWindow(list)
 	w.Resize(fyne.NewSize(200, 1000))
 	template := newListItem(fyne.NewContainerWithLayout(layout.NewHBoxLayout(), NewIcon(theme.DocumentIcon()), NewLabel("Template Object")), nil)
@@ -59,7 +59,7 @@ func TestList_Resize(t *testing.T) {
 }
 
 func TestList_OffsetChange(t *testing.T) {
-	list := createList()
+	list := createList(1000)
 	w := test.NewWindow(list)
 	w.Resize(fyne.NewSize(200, 1000))
 	template := newListItem(fyne.NewContainerWithLayout(layout.NewHBoxLayout(), NewIcon(theme.DocumentIcon()), NewLabel("Template Object")), nil)
@@ -91,7 +91,7 @@ func TestList_OffsetChange(t *testing.T) {
 }
 
 func TestList_Hover(t *testing.T) {
-	list := createList()
+	list := createList(1000)
 	children := test.WidgetRenderer(list).(*listRenderer).children
 
 	for i := 0; i < 2; i++ {
@@ -104,7 +104,7 @@ func TestList_Hover(t *testing.T) {
 }
 
 func TestList_Selection(t *testing.T) {
-	list := createList()
+	list := createList(1000)
 	children := test.WidgetRenderer(list).(*listRenderer).children
 
 	assert.Equal(t, children[0].(*listItem).statusIndicator.FillColor, theme.BackgroundColor())
@@ -118,7 +118,7 @@ func TestList_Selection(t *testing.T) {
 }
 
 func TestList_DataChange(t *testing.T) {
-	list := createList()
+	list := createList(1000)
 	w := test.NewWindow(list)
 	w.Resize(fyne.NewSize(200, 1000))
 	children := test.WidgetRenderer(list).(*listRenderer).children
@@ -133,7 +133,7 @@ func TestList_DataChange(t *testing.T) {
 }
 
 func TestList_ThemeChange(t *testing.T) {
-	list := createList()
+	list := createList(1000)
 	w := test.NewWindow(list)
 	w.Resize(fyne.NewSize(200, 1000))
 
@@ -146,9 +146,15 @@ func TestList_ThemeChange(t *testing.T) {
 	})
 }
 
-func createList() *List {
+func TestList_SmallList(t *testing.T) {
+	list := createList(1)
+	list.Refresh()
+	assert.Equal(t, list.Length(), 1)
+}
+
+func createList(items int) *List {
 	var data []string
-	for i := 0; i < 1000; i++ {
+	for i := 0; i < items; i++ {
 		data = append(data, fmt.Sprintf("Test Item %d", i))
 	}
 

--- a/widget/list_test.go
+++ b/widget/list_test.go
@@ -147,9 +147,24 @@ func TestList_ThemeChange(t *testing.T) {
 }
 
 func TestList_SmallList(t *testing.T) {
-	list := createList(1)
+	var data []string
+	data = append(data, "Test Item 0")
+
+	list := NewList(
+		func() int {
+			return len(data)
+		},
+		func() fyne.CanvasObject {
+			return fyne.NewContainerWithLayout(layout.NewHBoxLayout(), NewIcon(theme.DocumentIcon()), NewLabel("Template Object"))
+		},
+		func(index int, item fyne.CanvasObject) {
+			item.(*fyne.Container).Objects[1].(*Label).SetText(data[index])
+		},
+	)
+	list.Resize(fyne.NewSize(200, 1000))
+
+	data = append(data, "Test Item 1")
 	list.Refresh()
-	assert.Equal(t, list.Length(), 1)
 }
 
 func createList(items int) *List {

